tls/s2n_cipher_suites.c

```diff
@@ -1074,42 +1074,31 @@ int s2n_cipher_suites_cleanup(void)
     return 0;
 }
 
-struct s2n_cipher_suite *s2n_cipher_suite_from_wire(const uint8_t cipher_suite[S2N_TLS_CIPHER_SUITE_LEN])
-{
-    int low = 0;
-    int top = (sizeof(s2n_all_cipher_suites) / sizeof(struct s2n_cipher_suite *)) - 1;
-    /* Perform a textbook binary search */
-    while (low <= top) {
-        /* Check in the middle */
-        int mid = low + ((top - low) / 2);
-        int m = memcmp(s2n_all_cipher_suites[mid]->iana_value, cipher_suite, 2);
-
-        if (m == 0) {
-            return s2n_all_cipher_suites[mid];
-        } else if (m > 0) {
-            top = mid - 1;
-        } else if (m < 0) {
-            low = mid + 1;
-        }
-    }
-
-    return NULL;
-}
-
 int s2n_set_cipher_as_client(struct s2n_connection *conn, uint8_t wire[S2N_TLS_CIPHER_SUITE_LEN])
 {
     notnull_check(conn);
     notnull_check(conn->secure.cipher_suite);
-    struct s2n_cipher_suite *cipher_suite;
 
-    /* See if the cipher is one we support */
-    cipher_suite = s2n_cipher_suite_from_wire(wire);
+    const struct s2n_security_policy *security_policy;
+    GUARD(s2n_connection_get_security_policy(conn, &security_policy));
+    notnull_check(security_policy);
+
+    struct s2n_cipher_suite *cipher_suite = NULL;
+    for (size_t i = 0; i < security_policy->cipher_preferences->count; i++) {
+        const uint8_t *ours = security_policy->cipher_preferences->suites[i]->iana_value;
+        if (memcmp(wire, ours, S2N_TLS_CIPHER_SUITE_LEN) == 0) {
+            cipher_suite = security_policy->cipher_preferences->suites[i];
+            break;
+        }
+    }
     ENSURE_POSIX(cipher_suite != NULL, S2N_ERR_CIPHER_NOT_SUPPORTED);
+    ENSURE_POSIX(cipher_suite->available, S2N_ERR_CIPHER_NOT_SUPPORTED);
 
-    /* From RFC section: https://tools.ietf.org/html/rfc8446#section-4.2.11:
-     * Client MUST verify that the server selected a cipher suite indicating a Hash
-     * associated with the chosen PSK if it exists. 
-     * */
+    /** Clients MUST verify
+     *= https://tools.ietf.org/rfc/rfc8446#section-4.2.11
+     *# that the server selected a cipher suite
+     *# indicating a Hash associated with the PSK
```

Comment:
We need to include `associated with the chosen PSK if it exists` to the sentence.

Reply:
Unfortunately, not an option. You can't modify RFC snippets, or the tool can't recognize them.

I also think "the PSK" is clear enough in this context.